From c38359ca47adaa5fbf4146503119abe7fc1f0014 Mon Sep 17 00:00:00 2001
From: JimMoen <LnJimMoen@outlook.com>
Date: Mon, 22 Jun 2026 14:25:52 +0800
Subject: [PATCH] feature: allow disable hardcoded touchpad gestures

setting KWIN_WAYLAND_NO_GESTURE environment variable to disable the hardcoded touchpad gestures
---
 src/plugins/overview/overvieweffect.cpp | 26 ++++----
 src/virtualdesktops.cpp                 | 84 +++++++++++++------------
 2 files changed, 57 insertions(+), 53 deletions(-)

diff --git a/src/plugins/overview/overvieweffect.cpp b/src/plugins/overview/overvieweffect.cpp
index d9be30e..ac10c38 100644
--- a/src/plugins/overview/overvieweffect.cpp
+++ b/src/plugins/overview/overvieweffect.cpp
@@ -32,18 +32,20 @@ OverviewEffect::OverviewEffect()
     , m_gridBorder(new EffectTogglableTouchBorder(m_gridState))
     , m_shutdownTimer(new QTimer(this))
 {
-    auto gesture = new EffectTogglableGesture(m_overviewState);
-    gesture->addTouchpadSwipeGesture(SwipeDirection::Up, 4);
-    gesture->addTouchscreenSwipeGesture(SwipeDirection::Up, 3);
-
-    auto transitionGesture = new EffectTogglableGesture(m_transitionState);
-    transitionGesture->addTouchpadSwipeGesture(SwipeDirection::Up, 4);
-    transitionGesture->addTouchscreenSwipeGesture(SwipeDirection::Up, 3);
-    m_transitionState->stop();
-
-    auto gridGesture = new EffectTogglableGesture(m_gridState);
-    gridGesture->addTouchpadSwipeGesture(SwipeDirection::Down, 4);
-    gridGesture->addTouchscreenSwipeGesture(SwipeDirection::Down, 3);
+    if (!qEnvironmentVariableIsSet("KWIN_WAYLAND_NO_GESTURE")) {
+        auto gesture = new EffectTogglableGesture(m_overviewState);
+        gesture->addTouchpadSwipeGesture(SwipeDirection::Up, 4);
+        gesture->addTouchscreenSwipeGesture(SwipeDirection::Up, 3);
+
+        auto transitionGesture = new EffectTogglableGesture(m_transitionState);
+        transitionGesture->addTouchpadSwipeGesture(SwipeDirection::Up, 4);
+        transitionGesture->addTouchscreenSwipeGesture(SwipeDirection::Up, 3);
+        m_transitionState->stop();
+
+        auto gridGesture = new EffectTogglableGesture(m_gridState);
+        gridGesture->addTouchpadSwipeGesture(SwipeDirection::Down, 4);
+        gridGesture->addTouchscreenSwipeGesture(SwipeDirection::Down, 3);
+    }
 
     connect(m_overviewState, &EffectTogglableState::inProgressChanged, this, &OverviewEffect::overviewGestureInProgressChanged);
     connect(m_overviewState, &EffectTogglableState::partialActivationFactorChanged, this, &OverviewEffect::overviewPartialActivationFactorChanged);
diff --git a/src/virtualdesktops.cpp b/src/virtualdesktops.cpp
index a7414d2..b1120ff 100644
--- a/src/virtualdesktops.cpp
+++ b/src/virtualdesktops.cpp
@@ -777,50 +777,52 @@ void VirtualDesktopManager::initShortcuts()
 
     // Gestures
     // These connections decide which desktop to end on after gesture ends
-    connect(m_swipeGestureReleasedX.get(), &QAction::triggered, this, &VirtualDesktopManager::gestureReleasedX);
-    connect(m_swipeGestureReleasedY.get(), &QAction::triggered, this, &VirtualDesktopManager::gestureReleasedY);
-    const auto emitCurrentChanging = [this]() {
-        if (m_perOutputVirtualDesktops) {
-            LogicalOutput *output = workspace()->activeOutput();
-            Q_EMIT currentChanging(currentDesktop(output), m_currentDesktopOffset, output);
-        } else {
-            const auto outputs = workspace()->outputs();
-            for (LogicalOutput *output : outputs) {
+    if (!qEnvironmentVariableIsSet("KWIN_WAYLAND_NO_GESTURE")) {
+        connect(m_swipeGestureReleasedX.get(), &QAction::triggered, this, &VirtualDesktopManager::gestureReleasedX);
+        connect(m_swipeGestureReleasedY.get(), &QAction::triggered, this, &VirtualDesktopManager::gestureReleasedY);
+        const auto emitCurrentChanging = [this]() {
+            if (m_perOutputVirtualDesktops) {
+                LogicalOutput *output = workspace()->activeOutput();
                 Q_EMIT currentChanging(currentDesktop(output), m_currentDesktopOffset, output);
+            } else {
+                const auto outputs = workspace()->outputs();
+                for (LogicalOutput *output : outputs) {
+                    Q_EMIT currentChanging(currentDesktop(output), m_currentDesktopOffset, output);
+                }
             }
-        }
-    };
+        };
 
-    const auto left = [this, emitCurrentChanging](qreal cb) {
-        if (grid().width() > 1) {
-            m_currentDesktopOffset.setX(cb);
-            emitCurrentChanging();
-        }
-    };
-    const auto right = [this, emitCurrentChanging](qreal cb) {
-        if (grid().width() > 1) {
-            m_currentDesktopOffset.setX(-cb);
-            emitCurrentChanging();
-        }
-    };
-    input()->registerTouchpadSwipeShortcut(SwipeDirection::Left, 3, m_swipeGestureReleasedX.get(), left);
-    input()->registerTouchpadSwipeShortcut(SwipeDirection::Right, 3, m_swipeGestureReleasedX.get(), right);
-    input()->registerTouchpadSwipeShortcut(SwipeDirection::Left, 4, m_swipeGestureReleasedX.get(), left);
-    input()->registerTouchpadSwipeShortcut(SwipeDirection::Right, 4, m_swipeGestureReleasedX.get(), right);
-    input()->registerTouchpadSwipeShortcut(SwipeDirection::Down, 3, m_swipeGestureReleasedY.get(), [this, emitCurrentChanging](qreal cb) {
-        if (grid().height() > 1) {
-            m_currentDesktopOffset.setY(-cb);
-            emitCurrentChanging();
-        }
-    });
-    input()->registerTouchpadSwipeShortcut(SwipeDirection::Up, 3, m_swipeGestureReleasedY.get(), [this, emitCurrentChanging](qreal cb) {
-        if (grid().height() > 1) {
-            m_currentDesktopOffset.setY(cb);
-            emitCurrentChanging();
-        }
-    });
-    input()->registerTouchscreenSwipeShortcut(SwipeDirection::Left, 3, m_swipeGestureReleasedX.get(), left);
-    input()->registerTouchscreenSwipeShortcut(SwipeDirection::Right, 3, m_swipeGestureReleasedX.get(), right);
+        const auto left = [this, emitCurrentChanging](qreal cb) {
+            if (grid().width() > 1) {
+                m_currentDesktopOffset.setX(cb);
+                emitCurrentChanging();
+            }
+        };
+        const auto right = [this, emitCurrentChanging](qreal cb) {
+            if (grid().width() > 1) {
+                m_currentDesktopOffset.setX(-cb);
+                emitCurrentChanging();
+            }
+        };
+        input()->registerTouchpadSwipeShortcut(SwipeDirection::Left, 3, m_swipeGestureReleasedX.get(), left);
+        input()->registerTouchpadSwipeShortcut(SwipeDirection::Right, 3, m_swipeGestureReleasedX.get(), right);
+        input()->registerTouchpadSwipeShortcut(SwipeDirection::Left, 4, m_swipeGestureReleasedX.get(), left);
+        input()->registerTouchpadSwipeShortcut(SwipeDirection::Right, 4, m_swipeGestureReleasedX.get(), right);
+        input()->registerTouchpadSwipeShortcut(SwipeDirection::Down, 3, m_swipeGestureReleasedY.get(), [this, emitCurrentChanging](qreal cb) {
+            if (grid().height() > 1) {
+                m_currentDesktopOffset.setY(-cb);
+                emitCurrentChanging();
+            }
+        });
+        input()->registerTouchpadSwipeShortcut(SwipeDirection::Up, 3, m_swipeGestureReleasedY.get(), [this, emitCurrentChanging](qreal cb) {
+            if (grid().height() > 1) {
+                m_currentDesktopOffset.setY(cb);
+                emitCurrentChanging();
+            }
+        });
+        input()->registerTouchscreenSwipeShortcut(SwipeDirection::Left, 3, m_swipeGestureReleasedX.get(), left);
+        input()->registerTouchscreenSwipeShortcut(SwipeDirection::Right, 3, m_swipeGestureReleasedX.get(), right);
+    }
 
     // axis events
     input()->registerAxisShortcut(Qt::MetaModifier | Qt::AltModifier, PointerAxisDown,
-- 
2.54.0

